From 5e020b00a194eff90619952b46a8ea6c17f92340 Mon Sep 17 00:00:00 2001
From: Kalyan Alle <kalyan.alle@amd.com>
Date: Tue, 20 Mar 2018 12:08:46 +0530
Subject: [PATCH 3317/5855] drm/amd/display/amdgpu_dm:fixes the glmark2 hard
 hang

issue on 4.9kernel,POR stack amd-ese-por-4.9-2018_10

Change 1 & 2: Fixes the issue with dangling pointer cleanup

Change 2: Reverting the commit 86f4ec6899763302591c791bdb41ba136fc0f9a5
drm/amdgpu: Do not ref-count on validation set. This commit is from branch
amd-17.50

Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c | 10 +---------
 1 file changed, 1 insertion(+), 9 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
index f95e15c..9d367b0 100644
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
@@ -3173,11 +3173,6 @@ dm_drm_plane_duplicate_state(struct drm_plane *plane)
 
 	__drm_atomic_helper_plane_duplicate_state(plane, &dm_plane_state->base);
 
-	if (old_dm_plane_state->dc_state) {
-		dm_plane_state->dc_state = old_dm_plane_state->dc_state;
-		dc_plane_state_retain(dm_plane_state->dc_state);
-	}
-
 	return &dm_plane_state->base;
 }
 
@@ -3185,10 +3180,6 @@ void dm_drm_plane_destroy_state(struct drm_plane *plane,
 					   struct drm_plane_state *state)
 {
 	struct dm_plane_state *dm_plane_state = to_dm_plane_state(state);
-
-	if (dm_plane_state->dc_state)
-		dc_plane_state_release(dm_plane_state->dc_state);
-
 	drm_atomic_helper_plane_destroy_state(plane, state);
 }
 
@@ -4831,6 +4822,7 @@ int amdgpu_dm_atomic_check(struct drm_device *dev,
 		old_acrtc_state = to_dm_crtc_state(crtc->state);
 
 		if (old_acrtc_state->stream) {
+			dc_stream_retain(old_acrtc_state->stream);
 			set[set_count].stream = old_acrtc_state->stream;
 			++set_count;
 		}
-- 
2.7.4

